services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -0,0 +1,225 @@
+                                                                        Collection<FailedFileDownload> failedFileDownloads,
+                                                                        S3Object s3Object) {
+        String directory = downloadDirectoryRequest.destinationDirectory().toString();
+        String relativePath = getRelativePath(getDelimiter(downloadDirectoryRequest),
+                                              s3Object.key());
+
+        Path destinationPath = fileSystem.getPath(directory, relativePath);
+        DownloadFileRequest request = downloadFileRequest(downloadDirectoryRequest, s3Object, destinationPath);
+
+        try {
+            log.debug(() -> "Sending download request " + request);
+            createParentDirectoriesIfNeeded(destinationPath);
+
+            CompletableFuture<CompletedFileDownload> future = downloadFileFunction.apply(request).completionFuture();
+            future.whenComplete((r, t) -> {
+                if (t != null) {
+                    failedFileDownloads.add(FailedFileDownload.builder()
+                                                              .exception(t)
+                                                              .request(request)
+                                                              .build());
+                }
+            });
+            return future;
+        } catch (Throwable throwable) {
+            failedFileDownloads.add(FailedFileDownload.builder()
+                                                      .exception(throwable)
+                                                      .request(request)
+                                                      .build());
+            return CompletableFutureUtils.failedFuture(throwable);
+        }
+    }
+
+    private static DownloadFileRequest downloadFileRequest(DownloadDirectoryRequest downloadDirectoryRequest,
+                                                           S3Object s3Object,
+                                                           Path destinationPath) {
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder()
+                                                            .bucket(downloadDirectoryRequest.bucket())
+                                                            .key(s3Object.key())
+                                                            .build();
+        DownloadFileRequest.Builder requestBuilder = DownloadFileRequest.builder()
+                                                                        .destination(destinationPath)
+                                                                        .getObjectRequest(
+                                                                            getObjectRequest);
+
+
+        downloadDirectoryRequest.overrideConfiguration()
+                                .map(DownloadDirectoryOverrideConfiguration::downloadFileRequestTransformer)
```

Comment:
_\<Bit off topic but want to write down my thoughts so we can capture them somewhere\>_

Even though it's not an immediately planned feature/concern, I'm trying to think what it would look like to extend this download logic to simulate "sync" logic. E.g.,

> A s3 object will require copying if the sizes of the two s3 objects differ, the last modified time of the source is newer than the last modified time of the destination, or the s3 object does not exist under the specified bucket and prefix destination.

https://docs.aws.amazon.com/cli/latest/reference/s3/sync.html#examples

I'm not sure if the current `downloadFileRequestTransformer` is extensible enough to support something like this. Should we consider something like a `BiPredicate<GetObjectResponse, Path>` that allows users to declare if an individual download should be allowed to execute or not? If we supported something like that, then in theory, implementing sync-download logic would be as simple as implementing it to replicate the behavior quoted above.